### PR TITLE
Sam/deployment work

### PR DESCRIPTION
**Pull request by sam0044**
### Description

This PR aims to let the admin choose the deployment for each specialization. A defaultmodel field is set to have a model set for non specialization chats. This PR doesn't let the admin choose the embeddings just yet. But it does add the change where which  ever chat completion model you use, the embedding model for the same connection is used in chat request settings for azure. 

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [Contribution Guidelines](https://github.com/microsoft/chat-copilot/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/chat-copilot/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

**Comment by wdavidcamden** [comment]
I think we need to update the appsettings throughout the repo now.

File: helm/config/prod/appsettings.Production.json [hunk 1]
```diff
@@ -54,15 +50,33 @@
...
     "Enabled": true,
-    "AdminGroupMembershipId": "e1aae29c-d680-4e7e-a90a-d10cdb18fd74",
+    "DefaultModel": "gpt-4o",
     "DefaultSpecializationDescription": "This is a chat between an intelligent AI bot named Copilot and one or more participants. SK stands for Semantic Kernel, the AI platform used to build the bot.",
     "DefaultSpecializationImage": "https://quartech.com/wp-content/uploads/2021/09/Quartech_Logo_RGB.png",
-    "AzureConfig": {
-      "Endpoint": "https://searchs-ncus-qsl-openai-001.search.windows.net", // spellchecker:disable-line
-      "VectorizationSource": {
-        "Endpoint": "https://ncus-qsl-openai-poc.openai.azure.com/openai/deployments/text-embedding-ada-002/embeddings?api-version=2023-07-01-preview"
+    "AdminGroupMembershipId": "2dc97190-39a7-48b6-9b07-353c15100d5e",
```
Comment (wdavidcamden) [inline helm/config/prod/appsettings.Production.json:56]: this needs to be e1aae29c-d680-4e7e-a90a-d10cdb18fd74

File: helm/config/prod/appsettings.Production.json [hunk 2]
```diff
@@ -29,10 +29,6 @@
       "EmbeddingGeneratorTypes": ["AzureOpenAI"]
     },
     "Services": {
-      "AzureOpenAIText": {
-        "Deployment": "gpt-4o",
-        "Endpoint": "https://ncus-qsl-openai-poc.openai.azure.com/"
-      },
       "AzureOpenAIEmbedding": {
         "Deployment": "text-embedding-ada-002",
         "Endpoint": "https://ncus-qsl-openai-poc.openai.azure.com/"
```
Comment (wdavidcamden) [inline helm/config/prod/appsettings.Production.json:34]: can we remove this now?

File: helm/config/prod/appsettings.Production.json [hunk 3]
```diff
@@ -54,15 +50,33 @@
...
+    ],
+    "AISearchDeploymentConnections": [
+      {
+        "Name": "searchs-ncus-qsl-openai-001", // spellchecker:disable-line
+        "Endpoint": "https://searchs-ncus-qsl-openai-001.search.windows.net" // spellchecker:disable-line
+      }
+    ],
     "SpecializationIndexes": [
       {
         "IndexName": "david-probate-files",
```
Comment (wdavidcamden) [inline helm/config/prod/appsettings.Production.json:82]: need to add the search connection that is used for the index

File: helm/config/prod/appsettings.Production.json (hunk 2 above)
Comment (sam0044) [reply helm/config/prod/appsettings.Production.json:34]: No, not the embedding one.

File: helm/config/prod/appsettings.Production.json (hunk 1 above)
Comment (sam0044) [reply helm/config/prod/appsettings.Production.json]: Done

File: helm/config/prod/appsettings.Production.json (hunk 2 above)
Comment (wdavidcamden) [reply helm/config/prod/appsettings.Production.json:34]: can we just specify the default similar to the model?

File: helm/config/prod/appsettings.Production.json (hunk 3 above)
Comment (sam0044) [reply helm/config/prod/appsettings.Production.json:82]: done

File: helm/config/prod/appsettings.Production.json (hunk 2 above)
Comment (sam0044) [reply helm/config/prod/appsettings.Production.json:34]: I can look into that

File: helm/config/prod/appsettings.Production.json [hunk 4]
```diff
@@ -54,39 +50,63 @@
   },
   "QAzureOpenAIChatConfig": {
     "Enabled": true,
-    "AdminGroupMembershipId": "e1aae29c-d680-4e7e-a90a-d10cdb18fd74",
+    "DefaultModel": "gpt-4o",
```
Comment (wdavidcamden) [inline helm/config/prod/appsettings.Production.json:53]: add a Default Embedding Model

File: helm/config/prod/appsettings.Production.json [hunk 5]
```diff
@@ -54,39 +50,63 @@
   },
   "QAzureOpenAIChatConfig": {
     "Enabled": true,
```
Comment (wdavidcamden) [inline helm/config/prod/appsettings.Production.json:52]: I dont think this is used anymore ...

File: helm/config/prod/appsettings.Production.json (hunk 4 above)
Comment (sam0044) [reply helm/config/prod/appsettings.Production.json:57]: part of a seperate ticket

File: helm/config/prod/appsettings.Production.json (hunk 5 above)
Comment (sam0044) [reply helm/config/prod/appsettings.Production.json:56]: it seems to be being used within the code

Comment (wdavidcamden) [reply helm/config/prod/appsettings.Production.json:56]: to be addressed in another ticket

File: helm/config/dev/appsettings.Development.json [hunk 6]
```diff
@@ -54,43 +54,82 @@
...
-      "Endpoint": "https://searchs-ncus-qsl-openai-001.search.windows.net", // spellchecker:disable-line
-      "VectorizationSource": {
-        "Endpoint": "https://ncus-qsl-openai-poc.openai.azure.com/openai/deployments/text-embedding-ada-002/embeddings?api-version=2023-07-01-preview"
+    "AdminGroupMembershipId": "e1aae29c-d680-4e7e-a90a-d10cdb18fd74",
+    "OpenAIDeploymentConnections": [
+      {
+        "Name": "ncus-qsl-openai-poc",
+        "Endpoint": "https://ncus-qsl-openai-poc.openai.azure.com/",
+        "EmbeddingEndpoint": "https://ncus-qsl-openai-poc.openai.azure.com/openai/deployments/text-embedding-ada-002/embeddings?api-version=2023-05-15",
+        "ChatCompletionDeployments": ["gpt-4o", "gpt-4"],
```
[the remark this answers is not in the document]
Comment (hansonchun-quartech) [reply helm/config/dev/appsettings.Development.json:66]: Just talked to David about this, seems we need to set up some kind of user settings secret key? Could you create a section in the README on how to do that?

**Review by hansonchun-quartech — COMMENTED** [review]
It seems I'm missing something to in the set up to properly test this PR. I remember in your demo the "General" should be a drop down where you can select different specializations. However, I don't see any other specializations besides the General one. Do I need to set up some app settings to see them?

image

File: helm/config/dev/appsettings.Development.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (sam0044) [reply helm/config/dev/appsettings.Development.json:66]: The general dropdown is a separate feature that is not part of this feature so you won't see it in this branch. 
You need to create the specializations in the admin view and then also have the appropriate settings in the appsettings. I assumed that this PR would be in before you guys would get started so only david was the reviewer on this PR so he knew everything on how to test it. I can make a post in the general on teams with the appsettings required to boot the app

[the remark this answers is not in the document]
Comment (hansonchun-quartech) [reply helm/config/dev/appsettings.Development.json:66]: Yes that'd be great. Thanks

File: webapi/Plugins/Chat/ChatPlugin.cs [hunk 7]
```diff
@@ -799,8 +803,17 @@ private async Task<CopilotChatMessage> StreamResponseToClientAsync(
...
+        var specialization =
+            specializationkey == "general"
+                ? null
+                : await this._qSpecializationService.GetSpecializationAsync(specializationkey);
+        var serviceId =
+            (specialization == null || specialization.Deployment == defaultModel)
+                ? "default"
+                : specialization.Deployment;
+        var chatCompletion = provider.GetKeyedService<IChatCompletionService>(serviceId);
         var stream = chatCompletion.GetStreamingChatMessageContentsAsync(
```
Comment (hansonchun-quartech) [inline webapi/Plugins/Chat/ChatPlugin.cs:817]: Maybe put a null-conditional check on `chatCompletion` here

File: webapi/Controllers/SpecializationController.cs [hunk 8]
```diff
@@ -88,6 +88,19 @@ public List<string> GetAllSpecializationIndexes()
...
     }
 
+    /// <summary>
+    /// Get all chat completion deployments.
+    /// </summary>
+    /// <returns>A list of chat completion deployments.</returns>
+    [HttpGet]
+    [Route("specialization/deployments")]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status500InternalServerError)]
```
Comment (hansonchun-quartech) [inline webapi/Controllers/SpecializationController.cs:98]: Could put more ResponseTypes here (403, 404) to have more discerning error types

File: webapi/Plugins/Chat/Ext/QAzureOpenAIChatExtension.cs [hunk 9]
```diff
@@ -61,12 +64,30 @@ public bool isEnabled(string? specializationId)
...
-            var azureConfig = this._qAzureOpenAIChatOptions.AzureConfig;
+            var aiSearchDeploymentConnection =
+                this._qAzureOpenAIChatOptions.AISearchDeploymentConnections.FirstOrDefault(c =>
+                    c.Name == qSpecializationIndex.AISearchDeploymentConnection
+                );
+            var openAIDeploymentConnection = this._qAzureOpenAIChatOptions.OpenAIDeploymentConnections.FirstOrDefault(
+                c => c.Name == qSpecializationIndex.OpenAIDeploymentConnection
+            );
+            var EmbeddingEndpoint = this.GenerateEmbeddingEndpoint(
+                openAIDeploymentConnection.Endpoint,
```
Comment (hansonchun-quartech) [inline webapi/Plugins/Chat/Ext/QAzureOpenAIChatExtension.cs:82]: openAIDeploymentConnection may be null here

**Review by hansonchun-quartech — COMMENTED** [review]
Looks good. Just a few very minor things

**Comment by GrahamS-Quartech** [comment]
Once I got the appsettings setup correctly, I experienced no particular issues. I am able to enumerate deployments on the frontend and select each one successfully. I was able to ask the bot a question under each different deployment. General chat appears to still work correctly, seems that you have assigned it a default deployment.

**Review by GrahamS-Quartech — DISMISSED** [review]

File: webapi/Plugins/Chat/Ext/QAzureOpenAIChatExtension.cs (hunk 9 above)
Comment (sam0044) [reply webapi/Plugins/Chat/Ext/QAzureOpenAIChatExtension.cs:86]: done

File: webapi/Controllers/SpecializationController.cs (hunk 8 above)
Comment (sam0044) [reply webapi/Controllers/SpecializationController.cs:98]: done

File: webapi/Plugins/Chat/ChatPlugin.cs (hunk 7 above)
Comment (sam0044) [reply webapi/Plugins/Chat/ChatPlugin.cs:821]: done

**Review by hansonchun-quartech — APPROVED** [review]
https://github.com/user-attachments/assets/c4af5a31-7ff9-4f72-a682-0c06cf8acaa8

**Pull request merged**